src/main/java/io/cdap/plugin/spark/HierarchyFlattener.java

```diff
@@ -0,0 +1,507 @@
+    this.trueStr = config.getTrueValue();
+    this.falseStr = config.getFalseValue();
+    this.maxLevel = config.getMaxDepth();
+    this.parentChildMapping = config.getParentChildMapping();
+  }
+
+  /**
+   * Takes an input RDD and flattens it so that every possible ancestor to child relationship is present in the output.
+   * Each output record also is annotated with whether the ancestor is a root node, whether the child is a leaf node,
+   * and the distance from the ancestor to the child.
+   *
+   * Suppose the input data represents the hierarchy
+   *
+   *                      |--> 5
+   *             |--> 2 --|
+   *         1 --|        |--|
+   *             |--> 3      |--> 6
+   *                     4 --|
+   *
+   * This would be represented with the following rows:
+   *
+   *   [1 -> 1]
+   *   [1 -> 2]
+   *   [1 -> 3]
+   *   [2 -> 5]
+   *   [2 -> 6]
+   *   [4 -> 4]
+   *   [4 -> 6]
+   *
+   * and would generate the following output:
+   *
+   *   [1 -> 1, level:0, root:yes, leaf:no]
+   *   [1 -> 2, level:1, root:yes, leaf:no]
+   *   [1 -> 3, level:1, root:yes, leaf:no]
+   *   [1 -> 5, level:2, root:yes, leaf:yes]
+   *   [1 -> 6, level:2, root:yes, leaf:yes]
+   *   [2 -> 2, level:0, root:no, leaf:no]
+   *   [2 -> 5, level:1, root:no, leaf:yes]
+   *   [2 -> 6, level:1, root:no, leaf:yes]
+   *   [3 -> 3, level:0, root:no, leaf:yes]
+   *   [4 -> 4, level:0, root:yes, leaf:no]
+   *   [4 -> 6, level:1, root:yes, leaf:yes]
+   *   [5 -> 5, level:0, root:no, leaf:yes]
+   *   [6 -> 6, level:0, root:no, leaf:yes]
+   *
+   * @param context spark plugin context
+   * @param rdd input rdd representing a hierarchy
+   * @return flattened hierarchy with level, root, and leaf information
+   */
+  public JavaRDD<StructuredRecord> flatten(SparkExecutionPluginContext context, JavaRDD<StructuredRecord> rdd,
+                                           Schema outputSchema) {
+    Schema schema = context.getInputSchema();
+    SQLContext sqlContext = new SQLContext(context.getSparkContext());
+    StructType sparkSchema = DataFrames.toDataType(schema);
+    Dataset<Row> input = sqlContext.createDataFrame(rdd.map(record -> DataFrames.toRow(record, sparkSchema)).rdd(),
+                                                    sparkSchema);
+    // cache the input so that the previous stages don't get re-processed
+    input = input.persist(StorageLevel.DISK_ONLY());
+
+    // field names without the parent and child fields.
+    List<String> dataFieldNames = schema.getFields().stream()
+      .map(Schema.Field::getName)
+      .filter(name -> !name.equals(parentCol) && !name.equals(childCol))
+      .collect(Collectors.toList());
+
+    /*
+       The approach is to take N passes through the hierarchy, where N is the maximum depth of the tree.
+       In iteration i, paths of length i - 1 are found.
+
+       The first step is to generate the 0 distance paths.
+
+       With input rows:
+
+         [parent:1, child:1, category:grocery]
```

Comment:
nit: In the example here, id=1 has category=grocery. Elsewhere, id=1 has category=plants. It would be good to use the same example data everywhere.